spec/policies/task_file_policy_spec.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe TaskFilePolicy do
+  subject { described_class.new(user, task_file) }
+
+  let(:user) { nil }
+  let(:task_user) { create(:user) }
+  let(:groups) { [] }
+
+  let(:access_level) { :private }
+  let(:task) { create(:task, user: task_user, access_level:, groups:) }
+  let(:fileable) { task }
+  let(:task_file) { create(:task_file, fileable:) }
+
+  context 'with a user and private task' do
+    let(:user) { create(:user) }
+
+    it { is_expected.to forbid_all_actions }
+
+    context 'when user is admin' do
+      let(:user) { create(:admin) }
+
+      it { is_expected.to permit_all_actions }
+    end
+
+    context 'when task is from user' do
+      let(:task_user) { user }
+
+      it { is_expected.to permit_only_actions(%i[download_attachment extract_text_data]) }
+    end
+
+    context 'when task is in same group' do
+      let(:user) { create(:user) }
+      let(:group_memberships) { [build(:group_membership, :with_admin), build(:group_membership, user:, role: :confirmed_member)] }
+      let(:groups) { create_list(:group, 1, group_memberships:) }
+
+      it { is_expected.to permit_only_actions(%i[download_attachment extract_text_data]) }
+    end
+
+    context 'when task_file belongs to a test' do
+      let(:fileable) { create(:test, task:) }
+
+      it { is_expected.to forbid_all_actions }
+
+      context 'when task is from user' do
+        let(:task_user) { user }
+
+        it { is_expected.to permit_only_actions(%i[download_attachment extract_text_data]) }
+      end
+    end
+  end
+
+  context 'with a user and public task' do
+    let(:user) { create(:user) }
+    let(:access_level) { :public }
+
+    # why only permit downloading attachments when the text could be extracted after downloading anyway?
```

Comment:
Good question. Originally, `extract_text` is designed to be part of an update operation, thus requiring write access to the task (file). It got introduced by following the ProFormA specification, which differentiates between embedded and attached text (and binary) files. Therefore, this action is designed to convert attached text files to embedded text.

However, when digging deeper into the implementation, it seems like the actual action `extract_text` is just returning the file content for an AJAX operation (which, in turn, will use the response to set the file content in the frontend, followed by a regular "Update" operation triggered through the user). This, in turn, raises the question why extracting is not permitted, but downloading is. If you ask me, both operations might behave similar (in the current implementation, where extracting text is mainly a frontend operation).

Reply:
So should I keep it as is or allow extracting text for all users that can download the file?

Reply:
Let's allow extracting text for anyone who can download the file. When doing so, please add a short comment to that, explaining why extracting is allowed. Potentially, you can also link to my previous comment (but that's not required, if you find a well and short explanation 😉).